Prefix widget

crakezztrseg opened the pull request:
Introduces a prefix widgets for each time slot if defined. 

This pr  solves the issue, to define widgets to display like comma or colon before a time slot, which should be styled differently then a "activeWidget" or a "centerWidget". Using "activeWidget"  would solve the issue, but then you would have  the animation on the active Element, which does not match the design, i wanted to implement.


@dhiyaaulauliyaa if smth is missing feel free to comment or note it. Is my first PR for a package

crakezztrseg committed:
add new functionality

crakezztrseg committed:
prefix-widget

crakezztrseg committed:
prefix-widget

crakezztrseg committed:
remove incorrect value

crakezztrseg committed:
remove incorrect value

dhiyaaulauliyaa commented:
Hi there, sorry for the delayed response—Been difficult to find bandwidth to work on this 🙏

First off, thanks so much for this PR!😄 I really appreciate that you’ve:
- Added documentation
- Written unit tests
- Followed our existing code structure

Implementation‑wise, the Prefix widget functionality looks great, and I’m happy to approve the overall approach. 🤜🤛

Before we merge, though, could you help with a few tweaks?

- Fix the linter error:
![image](https://github.com/user-attachments/assets/c6320388-c341-4bf7-b8dc-a796b2ca71ef)
** For “redundant values” warning in the unit test file, can just suppress it by adding rule ignore on that file

- Simplify the DateTimePickerFlexCalculator API
The helper method for collecting dateTimeTypes is redundant since the same list is exposed by the DateTimePickerOption extension (check on dateTimeTypes method). You can update the calculateTotalFlex signature to take the dateTimeTypes as a parameter, like this:
```dart
static int calculateTotalFlex({
  required List<DateTimeType> dateTimeTypes,
  required DateTimePickerItemFlex itemFlex,
  required DateTimePickerPrefixFlex prefixFlex,
  required DateTimePickerPrefixWidget prefixWidget,
}) 
```
then when calling the calculateTotalFlex method, pass the dateTimeTypes from the DateTimePickerOption.

- Extract layout logic into a helper class
To make the widget code cleaner, let’s move the flex/spacing calculations out of lib/src/scroll_date_time_picker.dart into the DateTimePickerHelper. 
The code that needs to be moved:
```dart
 final hasPrefixWidget = widget.prefixWidget
        .hasTypeSpecificPrefixWidgets &&
    widget.prefixWidget.getPrefixWidget(type) !=
        null;

// move this
final currenttotalFlex = hasPrefixWidget
    ? widget.itemFlex.getFlex(type) +
        widget.prefixFlex.getFlex(type)
    : widget.itemFlex.getFlex(type);

// move this
final totalFlex = DateTimePickerFlexCalculator
    .calculateTotalFlex(
  dateFormat: _option.dateFormat,
  itemFlex: widget.itemFlex,
  prefixFlex: widget.prefixFlex,
  prefixWidget: widget.prefixWidget,
);

return Expanded(
  flex: currenttotalFlex,
  child: Row(
    children: [
      // Add a spacer with the same flex as the prefix widget if it exists
      if (hasPrefixWidget)
        SizedBox(
          width: constraints.maxWidth *
              (widget.prefixFlex.getFlex(type) /
                  totalFlex), // move this
        ),
      Expanded(
        child: widget.centerWidget
                .getCenterWidget(type) ??
            const SizedBox(),
      ),
    ],
  ),
);
```
You can add the needed class as a param to DateTimePickerHelper class:
```dart
const DateTimePickerHelper(
    this.option,
    this.itemFlex,
    this.prefixFlex,
  );

  final DateTimePickerOption option;
  final DateTimePickerItemFlex itemFlex;
  final DateTimePickerPrefixFlex prefixFlex;
```

then you can create the method to calculate needed values for the flex of prefix widget and the main widget such as total widget flex & prefix widget width


If you’re short on time and would like me to tackle any of these, just let me know which ones and I’m happy to help. Thanks again for your contribution! 😄

crakezztrseg commented:
@dhiyaaulauliyaa 

i will tackle this when i am finished with our release, then i have some time. Thanks for providing input.

dhiyaaulauliyaa committed:
chore 📌: Updates Flutter SDK version to 3.27.0

Updates the Flutter SDK version used by FVM to 3.27.0 in both the root and example directories. This ensures the project uses the latest features and improvements of the Flutter framework.

Removes the now redundant VS Code settings file from the project root, as the relevant settings are already defined in the example project's VS Code settings.

Additionally, the commit updates the .gitignore files to exclude the FVM version cache and Swift Package Manager files.

# Conflicts:
#	.vscode/settings.json

dhiyaaulauliyaa committed:
chore 📌: Updates example project structure to 3.27.0

Refactors the example Flutter project to align with the Flutter v3.27.0 project template.

This involves changes to the Android build configuration, iOS project settings, and Linux build setup, which includes migrating to Gradle 8.3, updating Kotlin version and iOS deployment target to 12.0, adding required queries for text processing in AndroidManifest, adding missing linux support files and modernizing the build system.

dhiyaaulauliyaa committed:
chore 📌: Updates intl dependency version constraint

Updates the intl package version constraint in pubspec.yaml
to allow versions greater than or equal to 0.19.0 but less
than 0.21.0.

This ensures compatibility with a wider range of intl versions
while avoiding potential breaking changes in future releases.

dhiyaaulauliyaa committed:
docs(changelog): 📝 update changelog

dhiyaaulauliyaa committed:
chore(pubspec): 🔖 bump version to 0.2.2

dhiyaaulauliyaa committed:
chore 🔥: remove example web from docs folder

Folder with name "docs" will violate pub.dev package folder naming regulation, so we will migrate the github pages implementation to gh-pages branch, and done automatically using github action

dhiyaaulauliyaa committed:
chore: 🚀  Updates Flutter version in CI

Updates the Flutter version used in the CI workflows to 3.27.0.
This ensures consistency and takes advantage of the latest features and improvements in the Flutter SDK.

dhiyaaulauliyaa committed:
ci: 🚀  Updates Dart SDK version on publish action

dhiyaaulauliyaa committed:
ci: 🚀  Create action for example web deployment

dhiyaaulauliyaa committed:
ci: 🚀  Update action for example web deployment

dhiyaaulauliyaa committed:
ci: 🚀  Implement example web deployment on publish action

dhiyaaulauliyaa committed:
ci: 🚀  Separate publish & web deploy action

crakezztrseg committed:
bump versions

crakezztrseg committed:
bump versions

crakezztrseg committed:
revert

crakezztrseg committed:
revert

crakezztrseg committed:
Merge remote-tracking branch 'origin/master' into prefix-widget

# Conflicts:
#	.vscode/settings.json
#	example/pubspec.lock

crakezztrseg committed:
Implement PR remarks

crakezztrseg committed:
Implement PR remarks

crakezztrseg committed:
Dart fix

crakezztrseg committed:
Dart fix

crakezztrseg committed:
add class docs to datetime picker helper class

crakezztrseg committed:
add class docs to datetime picker helper class

crakezztrseg committed:
revert add class docs to datetime picker helper class

crakezztrseg committed:
revert add class docs to datetime picker helper class

crakezztrseg committed:
add class docs to datetime picker helper class

crakezztrseg committed:
add class docs to datetime picker helper class

crakezztrseg committed:
Merge branch 'prefix-widget' of https://github.com/crakezztrseg/scroll_datetime_picker into prefix-widget

# Conflicts:
#	example/pubspec.lock

crakezztrseg commented:
@dhiyaaulauliyaa 

i updated the PR with the remarks you said. I tried to fix on my end the verification of signatures, but i could re sign all my commits.  

Feel free to just copy the changes and make a new PR. I unfortunately dont have enough time now, to tinker more around that, as its company time. 

Sorry!
